### ISSUE TITLE
Implicit Grant: Password

### ISSUE THREAD

**Issue by justpie** [FriendsOfSymfony/FOSOAuthServerBundle#450]
How would I use this bundle in an angularJs project? Implicit grants seem to not be supported. Is there a workaround that I could look into or would I be better off using another bundle that does support implicit?

Thanks.

**Comment by Spomky** [comment]
Related to #256, #266 and #282

**Comment by justpie** [comment]
Thanks for the reply Spomky. I have come across those posts but they were posted in 2014 so i wasn't sure if its been updated since.

**Comment by mangelsnc** [comment]
Hi, 

I'm in the same case, as @justpie anyone knows anything about this? I need to use it on a REST API and don't want to store my secret in the Javascript code.

What's the best way to protect the API if implicit grant type is not an option?

**Comment by chmielot** [comment]
+1

**Comment by chmielot** [comment]
@Spomky I have a REST-API and I want to publish a native mobile device app that is using it. The REST-API is basically public and anybody could use it. I need the implicit grant for the following flow:
App: /authorize?client_id=xyz&grant_type=token
REST-service: returns an access token in a JSON response.
There is a contract between the REST-Service and the app that the app must store the never expiring access token and send it for subsequent requests. This way the REST-service uses the token to identify a specific device where the app is running on. The user that is using the app is absolutely not involved in any of the authorisation process. It's the user's device that is doing everything automatically.

I know about security concerns and the drawbacks, but my question is: Can I achieve that by implementing a custom grant type for the FOSOAuthServerBundle? Any other bundle I tried doesn't implement what I'm trying to do.

**Comment by Spomky** [comment]
> The REST-API is basically public and anybody could use it.

In this case there is no need to issue tokens.

> This way the REST-service uses the token to identify a specific device where the app is running on. The user that is using the app is absolutely not involved in any of the authorisation process.

With the Client Credentials Grant Type, OAuth2 designed for that purpose, but it is like killing a fly with a sledgehammer. What I suggest you is to simply use Json Web Tokens (JWT) instead of this bundle.
The app will generate a JWS (signed with a key shared between the app and the server) and send it on each API request. The Resource server just have to check the JWT and allow or not the access on the resource.

**Comment by chmielot** [comment]
> With the Client Credentials Grant Type, OAuth2 designed for that purpose, but it is like killing a fly with a sledgehammer

I don't fully agree. It is everywhere discouraged to save the client secret in a native mobile app which is what pointed me to the implicit grant type. On the other hand it is clear that it doesn't matter. The client secret can be treated just like the client_id which is "public" so the flow fits as well. If I insist on using Oauth2 for my purpose and I would go for the Client Credentials Grant Type I have one problem left when using this bundle: It's not possible to set the expiry of the token to "never" which is definitely a MUST for my use case (token == device identifier).

Why would I insist on using Oauth2? Because:
1. I am planning to provide a user login in the future to identify the concrete user and not just the device. Sticking to Oauth2 from the beginning seems reasonable then.
2. I want to implement a server side app which makes use of the API as well, here the Client Credentials Grant Type is the perfect match because the client secret is actually secret. The token is still expiring but I can use the refresh token flow or just request a new token in that case, because I am not using the token to identify a device.

@Spomky Knowing the context and the big picture now would you still recommend 
1. JWS and add Oauth2 for user login and server app API consumer later or 
2. would you implement the Client Credentials Grant Type from the beginning?
If the latter, can you point me to the right direction to remove the expiry from the token? Is the solution to implement a custom grant type anyway?

Thanks for suggesting JWS btw. I'm going to have a look at that now...

**Comment by Spomky** [comment]
You are right, the use of confidential clients on native mobile applications is discouraged [because of possible threats](https://tools.ietf.org/html/rfc6819#section-4.1). However, if you implement countermeasures then nothing prevent you from using those clients in such context.

According to the explanation you provided earlier, then it appears that OAuth2 could be interesting as you do not need just client authentication, but also user permission/authentication.
As per the RFC below, possible countermeasures are:

* Secure token storage (encrypted folder/disk).
* Lock(s) to prevent unauthorized device/application access
* Possibility to revoke the tokens
* Rotation of the tokens
* Rotation of the client credentials
